### PR TITLE
fix(api): nvim_set_current_dir doesn't set previous directory

### LINKED ISSUE #15757: `nvim_set_current_dir` does inform `:cd -`
### Neovim version (nvim -v)

NVIM v0.(5|6).0

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

both MacOS 11.5 and Ubuntu 20.04

### Terminal name/version

Iterm2

### $TERM environment variable

xterm-256color

### Installation

homebrew / unstable repo

### How to reproduce the issue

`:lua vim.api.nvim_set_current_dir("/")`

### Expected behavior

`:cd -` should bring us back to previous work dir?

not sure if this is meant to be this way by design, please close this issue if this is.

### Actual behavior

Errors out with

`E186: No previous directory`

### PR DESCRIPTION
**Pull request by smolck**
Fixes #15757.

Still need to add a test, but is this a good way of doing this? 

Also can add an additional argument allowing to disable this feature as was mentioned in the issue, just figure I'd hold off first to make sure that's wanted.

### CODE REVIEW

File: src/nvim/ex_docmd.c [hunk 1]
```diff
@@ -7710,8 +7710,12 @@ void free_cd_dir(void)
 /// Deal with the side effects of changing the current directory.
 ///
 /// @param scope  Scope of the function call (global, tab or window).
-void post_chdir(CdScope scope, bool trigger_dirchanged)
+void post_chdir(CdScope scope, bool trigger_dirchanged, char_u *previous_dir)
 {
+  if (previous_dir != NULL) {
+    prev_dir = vim_strsave(previous_dir);
```
Comment (muniter) [inline src/nvim/ex_docmd.c:7716]: I believe this is a leak, you need to free the old prev_dir.

https://github.com/neovim/neovim/blob/70b854dbd2ae9763f8061052e1074cb14be27c00/src/nvim/ex_docmd.c#L7787

**Comment by justinmk** [comment]
I would consider changing the impl of nvim_set_current_dir to just `nvim_exec("cd "+dir)`. This is not a hotpath, and `:cd` has no quoting rules.

**Comment by jamessan** [comment]
> This is not a hotpath, and `:cd` has no quoting rules.

`:cd` honors `cmdline-special`, so it certainly does need escaping.

**Comment by smolck** [comment]
> I would consider changing the impl of nvim_set_current_dir to just nvim_exec("cd "+dir). This is not a hotpath, and :cd has no quoting rules.

Done (although I'm not sure if the `concat_str` thing is the right/best way to do the string concatenating?). And apologies for the noob question, but not sure on the string escaping; I see `vim_strsave_escaped` is a thing, but assuming that's the right function, what to pass for the escape chars? `"\"\\"`?

**Comment by jamessan** [comment]
`vim_strsave_fnameescape` is the one you're looking for.

**Review by justinmk — DISMISSED** [review]

**Comment by justinmk** [comment]
oh needs a test

**Comment by smolck** [comment]
> oh needs a test

Done, although I don't test the string escaping, should probably do that?

File: test/functional/api/vim_spec.lua [hunk 2]
```diff
@@ -1,7 +1,9 @@
 local helpers = require('test.functional.helpers')(after_each)
 local Screen = require('test.functional.ui.screen')
+local lfs = require('lfs')
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua:3]: lfs is never needed, we should add a lint check for this. Nvim functions are available from `funcs` (line 13 below)

File: test/functional/api/vim_spec.lua [hunk 3]
```diff
@@ -487,6 +489,31 @@ describe('API', function()
     end)
   end)
 
+  describe('nvim_set_current_dir', function()
+    local start_dir
+
+    before_each(function()
+      clear()
+      lfs.mkdir("testdir")
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua:497]: typical pattern for test files is to prefix with "X". 
```suggestion
      funcs.mkdir("Xtestdir")
```

File: test/functional/api/vim_spec.lua [hunk 4]
```diff
@@ -487,6 +489,31 @@ describe('API', function()
...
+      start_dir = getcwd()
```
Comment (justinmk) [inline test/functional/api/vim_spec.lua:498]:
```suggestion
      start_dir = funcs.getcwd()
```

File: src/nvim/api/vim.c [hunk 5]
```diff
@@ -812,17 +813,9 @@ void nvim_set_current_dir(String dir, Error *err)
...
-    if (!try_end(err)) {
-      api_set_error(err, kErrorTypeException, "Failed to change directory");
-    }
-    return;
-  }
-
-  post_chdir(kCdScopeGlobal, true);
-  try_end(err);
+  char_u *cmd = concat_str((const char_u *)"cd ", (const char_u *)string);
+  nvim_exec(cstr_to_string(vim_strsave_fnameescape((const char *)cmd, false)), false, err);
```
Comment (justinmk) [inline src/nvim/api/vim.c:817]: the `vim_strsave_fnameescape` result needs to be freed.

asan failure:


```
2021-09-26T23:57:07.6046138Z = Direct leak of 13 byte(s) in 1 object(s) allocated from:
2021-09-26T23:57:07.6046709Z =     #0 0x4d8e9d in malloc ??:?
2021-09-26T23:57:07.6047354Z =     #1 0x10ff222 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:74:15
2021-09-26T23:57:07.6048162Z =     #2 0x10ff41c in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:108:15
2021-09-26T23:57:07.6049060Z =     #3 0x188b7ef in vim_strsave_escaped_ext /home/runner/work/neovim/neovim/src/nvim/strings.c:108:28
2021-09-26T23:57:07.6049996Z =     #4 0x188b42b in vim_strsave_escaped /home/runner/work/neovim/neovim/src/nvim/strings.c:79:10
2021-09-26T23:57:07.6050944Z =     #5 0xd2889f in vim_strsave_fnameescape /home/runner/work/neovim/neovim/src/nvim/ex_getln.c:4131:13
2021-09-26T23:57:07.6051864Z =     #6 0x6d2517 in nvim_set_current_dir /home/runner/work/neovim/neovim/src/nvim/api/vim.c:817:28
2021-09-26T23:57:07.6054155Z =     #7 0x5d54a2 in handle_nvim_set_current_dir /home/runner/work/neovim/neovim/build/src/nvim/auto/api/private/dispatch_wrappers.generated.h:4050:3

Comment (gpanders) [reply src/nvim/api/vim.c:817]: `cstr_to_string` also calls `xmemdupz` so there are two strings that need to be freed. Probably should just use `cstr_as_string` instead.

File: test/functional/api/vim_spec.lua [hunk 6]
```diff
@@ -487,6 +487,31 @@ describe('API', function()
...
+      start_dir = funcs.getcwd()
+    end)
+
+    after_each(function()
+      helpers.rmdir("Xtestdir")
+    end)
+
+    it('works', function()
+      meths.set_current_dir("Xtestdir")
+      eq(funcs.getcwd(), start_dir .. "/Xtestdir")
```
[the remark this answers is not in the document]
Comment (muniter) [reply test/functional/api/vim_spec.lua:505]: https://github.com/neovim/neovim/blob/e06936125a628d8a61069f5fbce68394292db4df/test/functional/helpers.lua#L788-L790

**Comment by zeertzjq** [comment]
If #15952 is merged you can use `changedir_func` instead.

**Pull request closed without merging**